feat: Add comprehensive security access control system

antigenius0910 opened the pull request:
# 🛡️ Add Comprehensive Security Access Control System

## Overview

This PR introduces a robust security access control system for the Slack MCP client, enabling organizations to control who can interact with the bot and in which channels. The system provides enterprise-grade security features while maintaining full backward compatibility.

## 🚀 Key Features

### Security Access Control

- __User-based access control__ - Whitelist specific Slack users
- __Channel-based access control__ - Restrict bot access to specific channels
- __Admin user privileges__ - Special users who bypass channel restrictions
- __Flexible vs Strict modes__ - Choose between OR vs AND logic for access control
- __Polite rejection messages__ - Customizable messages for unauthorized users
- __Comprehensive audit logging__ - Track all access attempts with detailed context

### Configuration Management

- __Environment-based configuration__ - All settings via environment variables
- __JSON configuration support__ - Structured configuration with validation
- __Sensible defaults__ - Security disabled by default for easy adoption
- __Runtime configuration updates__ - Dynamic security config changes

## 🔧 Environment Variables

```bash
# Basic Security Settings
SECURITY_ENABLED="false"                    # Enable/disable security
SECURITY_STRICT_MODE="false"               # Require both user AND channel whitelisting
SECURITY_LOG_UNAUTHORIZED="true"           # Log unauthorized access attempts

# Access Control Lists
SECURITY_ALLOWED_USERS="U1234567890,U0987654321"     # Comma-separated user IDs
SECURITY_ALLOWED_CHANNELS="C1234567890,C0987654321"  # Comma-separated channel IDs
SECURITY_ADMIN_USERS="U1111111111"                   # Admin users (bypass channel restrictions)

# Customization
SECURITY_REJECTION_MESSAGE="I'm sorry, but I don't have permission to respond in this context. Please contact your administrator if you believe this is an error."
```

## 🏗️ Architecture

### Access Modes

__Flexible Mode (Default)__ - `SECURITY_STRICT_MODE=false`

- Access granted if user OR channel is whitelisted
- Admin users always have access regardless of channel restrictions

__Strict Mode__ - `SECURITY_STRICT_MODE=true`

- Access requires BOTH user AND channel to be whitelisted
- Admin users still bypass channel restrictions

### Security Flow

1. __Message Received__ → Security check performed
2. __Access Denied__ → Polite rejection message sent + logged
3. __Access Granted__ → Message processed normally + logged

## 📁 Files Added/Modified

### New Files

- `internal/security/config.go` - Security configuration struct and parsing
- `internal/security/access_controller.go` - Core access control logic
- `internal/security/loader.go` - Configuration loading utilities
- `internal/config/security_config.go` - Security config integration
- `internal/slack/secure_client.go` - Security-enabled Slack client
- `docs/security.md` - Comprehensive security documentation

### Modified Files

- `cmd/main.go` - Integration of security-enabled client selection
- `internal/slack/client.go` - __Bug fix__: Fixed GetUserInfo timing issue
- `.env.example` - Added all security environment variables

## 🐛 Bug Fixes

Fixed a critical bug in the regular Slack client where `GetUserInfo()` was being called for all message events (including bot messages, edited messages, etc.) before validation checks, causing `user_not_found` errors with empty user IDs.

## ✅ Testing

- ✅ Security system successfully tested with real Slack integration
- ✅ Access control logs showing proper user/channel validation
- ✅ Bug fix verified - no more `user_not_found` errors
- ✅ Backward compatibility confirmed - existing functionality unchanged

## 🔒 Security Considerations

- __Zero Breaking Changes__ - Security disabled by default
- __Fail-Safe Design__ - Errors default to denying access
- __Audit Trail__ - All access attempts logged with context
- __Professional UX__ - Polite rejection messages don't reveal system details

## 📚 Documentation

Complete documentation added in `docs/security.md` including:

- Configuration examples
- Access mode explanations
- Troubleshooting guide
- Best practices
- How to get Slack user/channel IDs

## 🚀 Usage Example

```bash
# Enable security with flexible mode
SECURITY_ENABLED=true
SECURITY_STRICT_MODE=false

# Allow specific users and channels
SECURITY_ALLOWED_USERS=U1234567890,U0987654321
SECURITY_ALLOWED_CHANNELS=C1234567890,C0987654321

# Set admin users
SECURITY_ADMIN_USERS=U1111111111

# Custom rejection message
SECURITY_REJECTION_MESSAGE="Access denied. Please contact your IT administrator."
```

## 🎯 Benefits

- __Enterprise Ready__ - Production-grade access control for corporate environments
- __Flexible Deployment__ - Works in both permissive and restrictive environments
- __Operational Visibility__ - Comprehensive logging for security monitoring
- __User Friendly__ - Polite rejection messages maintain professional experience
- __Zero Disruption__ - Existing deployments continue working unchanged

tuannvm commented:
Thanks @antigenius0910 for the contribution. We have recently made a major change in how we structure the configuration. Please have a look at [docs/configuration.md](https://github.com/tuannvm/slack-mcp-client/blob/main/docs/configuration.md) and update accordingly.

Ideally, the configuration should look like this:

```
  {
    "slack": { ... },
    "mcp_servers": { ... },
    "llm_providers": { ... },
    "security": {
      "enabled": false,
      "strict_mode": false,
      "allowed_users": ["U1234567890"],
      "allowed_channels": ["C1234567890"],
      "admin_users": ["U1111111111"],
      "rejection_message": "...",
      "log_unauthorized": true
    }
  }
```

  This would require refactoring the security package to:
  1. Remove LoadSecurityConfig() and environment-only loading
  2. Add security config to your main Config struct
  3. Use existing environment override mechanism
  4. Remove the ConfigWithSecurity wrapper

tuannvm commented on internal/security/access_controller.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+// NewAccessController creates a new AccessController instance
+func NewAccessController(config *SecurityConfig, logger *logging.Logger) *AccessController {
```
We should use proper error handling instead of panics.
  ```
  func NewAccessController(config *SecurityConfig, logger *logging.Logger) (*AccessController, error) {
      if config == nil {
          return nil, fmt.Errorf("security config cannot be nil")
      }
      if logger == nil {
          return nil, fmt.Errorf("logger cannot be nil")
      }
      return &AccessController{config: config, logger: logger}, nil
  }
  ```

antigenius0910 commented on internal/slack/secure_client.go:
```diff
@@ -0,0 +1,171 @@
+// Package slackbot provides security-enabled Slack client functionality
+package slackbot
```
all the other file use the same package. don't think this is correct

antigenius0910 commented on cmd/main.go:
```diff
@@ -617,21 +617,39 @@ func startSlackClient(logger *logging.Logger, mcpClients map[string]*mcp.Client,
 		logger.Fatal("Failed to initialize Slack client: %v", err)
 	}
 
```
Configuration Loading Separated

antigenius0910 commented on cmd/main.go:
```diff
@@ -617,21 +617,39 @@ func startSlackClient(logger *logging.Logger, mcpClients map[string]*mcp.Client,
-	if err != nil {
-		logger.Fatal("Failed to initialize Slack client: %v", err)
+	// Use security-enabled client if security is enabled, otherwise use regular client
+	var slackClient interface{ Run() error }
```
Type safety is now fully restored with clean separation of concerns.

antigenius0910 commented:
Review Feedback Addressed

Thank you @tuannvm for the detailed and constructive feedback! we have implemented your suggestions in commit d25fb47:

### Key Fixes Applied:
- **Thread Safety**: Added `sync.RWMutex` to AccessController as suggested
- **Type Safety**: Implemented SlackRunner interface and factory pattern
- **Architecture**: Extracted client factory and simplified startup logic  
- **ThreadTimeStamp**: Added fallback to TimeStamp for non-threaded messages
- **Logging**: Used proper InfoKV/WarnKV instead of manual level field

### Live Testing Confirmed:
The security system is now fully functional with proper access control enforcement. All architectural improvements should working as expected.

Ready for review

tuannvm commented on internal/slack/slack_client.go:
```diff
@@ -0,0 +1,12 @@
+
+import "github.com/tuannvm/slack-mcp-client/internal/security"
+
+// SlackRunner defines the interface for a Slack client that can run and handle security
```
I'm not so sure we need this interface. Security should be a feature we can toggle on and off, rather than a separate implementation for secured Slack.
